xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,203 @@
+    }
+  }
+
+  /**
+   * The evaluate function performs the core authorization mechanism of CEL Evaluation Engine.
+   * Determines whether a gRPC call is allowed, denied, or unable to decide.
+   * @param args evaluate argument that is used to evaluate the RBAC conditions.
+   * @return an AuthorizationDecision generated by CEL Evaluation Engine.
+   * @throws InterpreterException if something goes wrong in CEL library.
+   */
+  public AuthorizationDecision evaluate(EvaluateArgs<ReqT, RespT> args) 
+      throws IllegalArgumentException {
+    if (this.rbacEngines.size() < 1 || this.rbacEngines.size() > 2) {
+      throw new IllegalArgumentException(
+        "Invalid RBAC list size, must provide either one RBAC or two RBACs. ");
+    } 
+    if (this.rbacEngines.size() == 2 && (this.rbacEngines.get(0).action != Action.DENY 
+        || this.rbacEngines.get(1).action != Action.ALLOW)) {
+      throw new IllegalArgumentException( "Invalid RBAC list, " 
+          + "must provide a RBAC with DENY action followed by a RBAC with ALLOW action. ");
+    }
+    AuthorizationDecision.Decision authorizationDecision = null;
+    List<String> matchingPolicyNames = new ArrayList<>();
+    List<String> unknownPolicyNames = new ArrayList<>();
+    // Set up activation used in CEL library's eval function.
+    Activation activation = Activation.copyOf(extractFields(args));
+    // Go through each RBAC in the Envoy RBAC list.
+    for (RbacEngine rbacEngine : rbacEngines) {
+      // Go through each condition in the RBAC policy.
+      for (Map.Entry<String, Expr> condition : rbacEngine.conditions.entrySet()) {
+        try {
+          if (matches(condition.getValue(), activation)) {
+            authorizationDecision = rbacEngine.action == Action.ALLOW 
+                ? AuthorizationDecision.Decision.ALLOW : AuthorizationDecision.Decision.DENY;
+            matchingPolicyNames.add(condition.getKey());
+          }
+        } catch (InterpreterException e) {
+          unknownPolicyNames.add(condition.getKey());
+        }
+      }
+      if (matchingPolicyNames.size() > 0) {
+        return new AuthorizationDecision(authorizationDecision, matchingPolicyNames);
+      }
+      if (unknownPolicyNames.size() > 0) {
+        return new AuthorizationDecision(
+            AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+      }
+    }
+    // No RBAC conditions matched and didn't find unknown conditions.
+    if (this.rbacEngines.size() == 1 && this.rbacEngines.get(0).action == Action.DENY) {
+      return new AuthorizationDecision(
+          AuthorizationDecision.Decision.ALLOW, new ArrayList<String>());
+    }
+    return new AuthorizationDecision(AuthorizationDecision.Decision.DENY, new ArrayList<String>());
+  }
+
+  /** Evaluate if a condition matches the given Enovy Attributes using CEL library. */
+  protected boolean matches(Expr condition, Activation activation) throws InterpreterException {
```

Comment:
if `matches` is only used once, can we move it to the evaluate logic? It might simplify the logic to handle IncompleteData.

Reply:
I also separated this `matches` function because in the unit test, I have to mock the behavior of `matches` in order to test the behavior of `evaluate` function.

Reply:
Sure, thanks for pointing it out tho!

Reply:
`protected` is for subclasses and AFAICS there aren't any. May be package-private vs private is the issue? Also my personal preference is to split code into multiple functions for modularity/testability.

Reply:
`matches` can be static since it doesn't use any instance vars of `AuthorizationEngine`?

Reply:
The function `matches` will be changed to a private function in the future.